src/python/pants/backend/nfpm/native_libs/elfdeps/analyze.py

```diff
@@ -0,0 +1,103 @@
+import json
+import sys
+import zipfile
+from collections.abc import Generator, Iterable
+from dataclasses import asdict, dataclass
+from pathlib import Path
+
+# elfdeps 0.2.0 added analyze_zipfile
+from elfdeps import ELFAnalyzeSettings, ELFInfo, SOInfo, analyze_dirtree, analyze_zipfile
+
+
+@dataclass(frozen=True)
+class ELFInfoAnalysis:
+    provides: tuple[SOInfo, ...]
+    requires: tuple[SOInfo, ...]
+
+    def __init__(self, provides: Iterable[SOInfo], requires: Iterable[SOInfo]):
+        object.__setattr__(self, "provides", tuple(sorted(provides)))
+        object.__setattr__(self, "requires", tuple(sorted(requires)))
+
+    def to_dict(self) -> dict[str, list[dict[str, str]]]:
+        # so_info: SOInfo(soname: str, version: str, marker: str)
+        # marker is one of "(64bit)" or ""
+        # str(so_info) = f"{soname}({version}){marker}"
+
+        def so_infos_to_dicts(so_infos: tuple[SOInfo, ...]) -> list[dict[str, str]]:
+            return [asdict(so_info) | {"so_info": str(so_info)} for so_info in so_infos]
+
+        return {
+            "provides": so_infos_to_dicts(self.provides),
+            "requires": so_infos_to_dicts(self.requires),
+        }
+
+    def to_json(self, indent=None, separators=(",", ":")) -> str:
+        return json.dumps(self.to_dict(), indent=indent, separators=separators)
+
+    @classmethod
+    def from_elf_infos(cls, elf_infos: Iterable[ELFInfo]) -> ELFInfoAnalysis:
+        provides: set[SOInfo] = set()
+        requires: set[SOInfo] = set()
+        for elf_info in elf_infos:
+            provides.update(elf_info.provides)  # elf_info.provides: list[SOInfo]
+            requires.update(elf_info.requires)  # elf_info.requires: list[SOInfo]
+
+        return cls(tuple(provides), tuple(requires))
+
+
+def analyze_wheel(wheel_path: Path, settings: ELFAnalyzeSettings) -> Generator[ELFInfo]:
+    print(".", end="", file=sys.stderr)  # a progress indicator
+    with zipfile.ZipFile(wheel_path, mode="r") as wheel:
+        yield from analyze_zipfile(wheel, settings=settings)
+
+
+def analyze_wheels_repo(wheel_repo: Path) -> ELFInfoAnalysis:
+    settings = ELFAnalyzeSettings(unique=True)
+
+    print(f"Analyzing wheels in {wheel_repo}", file=sys.stderr)
+    elf_infos: list[ELFInfo] = [
+        elf_info for wheel in wheel_repo.iterdir() for elf_info in analyze_wheel(wheel, settings)
+    ]
+    print(f"Done analyzing wheels in {wheel_repo}.", file=sys.stderr)  # end progress indicators
+
+    return ELFInfoAnalysis.from_elf_infos(elf_infos)
+
+
+def analyze_directory(directory: Path) -> ELFInfoAnalysis:
+    settings = ELFAnalyzeSettings(unique=True)
+
+    print(f"Analyzing files in {directory}", file=sys.stderr)
+    elf_infos: list[ELFInfo] = list(analyze_dirtree(directory, settings=settings))
+    print(f"Done analyzing files in {directory}.", file=sys.stderr)  # end progress indicators
+
+    return ELFInfoAnalysis.from_elf_infos(elf_infos)
+
+
+def main(args: list[str]) -> int:
```

Comment:
Probably good to have a big fat comment at the top of the file explaining that this code is intended to be run as a script in a subprocess, and is not intended to be imported and used in-process by other Pants code.

Reply:
Excellent suggestion. I'll add that as soon as I can (possibly on Monday).